src/Microsoft.Azure.WebJobs.Host/Scale/IScaleMonitor.cs

```diff
@@ -0,0 +1,45 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.WebJobs.Host.Scale
+{
+    /// <summary>
+    /// Interface defining a monitor that can participate in Azure Functions scaling decisions by
+    /// taking metrics samples, and scaling based on those samples.
+    /// </summary>
+    public interface IScaleMonitor
+    {
+        /// <summary>
+        /// Returns the unique ID for this monitor.
+        /// </summary>
+        /// <remarks>
+        /// This should be constant. It is used to correlate persisted metrics samples
+        /// with their corresponding monitor instance. E.g. for a QueueTrigger, this might
+        /// be of the form "{FunctionId}-QueueTrigger-{QueueName}".
+        /// </remarks>
+        string Id { get; }
+
+        /// <summary>
+        /// Return a current metrics sample by querying the event source.
+        /// </summary>
+        /// <returns>The <see cref="ScaleMetrics"/> sample.</returns>
+        Task<ScaleMetrics> GetMetricsAsync();
+
+        /// <summary>
+        /// Return the current scale status based on the specified context.
+        /// </summary>
+        /// <param name="context">The <see cref="ScaleStatusContext"/> to use to determine
+        /// the scale status.</param>
+        /// <returns>The scale status.</returns>
+        ScaleStatus GetScaleStatus(ScaleStatusContext context);
+    }
+
+    public interface IScaleMonitor<TMetrics> : IScaleMonitor where TMetrics : ScaleMetrics
+    {
+        new Task<TMetrics> GetMetricsAsync();
```

Comment:
The non-generic/generic split in these interfaces is somewhat unfortunate for implementors, though necessary I think. The functions host needs to interact with collections of IScaleMonitors (non-generic), however it also needs access to the actual TMetrics type, for serialization/deserialization, which it does at runtime. Open to other ideas on how to handle this.